forester/src/batched_ops.rs

```diff
@@ -0,0 +1,421 @@
+use std::sync::Arc;
+
+use borsh::BorshSerialize;
+use forester_utils::indexer::Indexer;
+use light_batched_merkle_tree::{
+    constants::DEFAULT_BATCH_STATE_TREE_HEIGHT,
+    event::{BatchAppendEvent, BatchNullifyEvent},
+    merkle_tree::{
+        AppendBatchProofInputsIx, BatchProofInputsIx, BatchedMerkleTreeAccount,
+        InstructionDataBatchAppendInputs, InstructionDataBatchNullifyInputs,
+    },
+    queue::BatchedQueueAccount,
+};
+use light_client::{rpc::RpcConnection, rpc_pool::SolanaRpcPool};
+use light_hasher::{Hasher, Poseidon};
+use light_prover_client::{
+    batch_append_with_proofs::get_batch_append_with_proofs_inputs,
+    batch_update::get_batch_update_inputs,
+    gnark::{
+        batch_append_with_proofs_json_formatter::BatchAppendWithProofsInputsJson,
+        batch_update_json_formatter::update_inputs_string,
+        constants::{PROVE_PATH, SERVER_ADDRESS},
+        proof_helpers::{compress_proof, deserialize_gnark_proof_json, proof_from_json_struct},
+    },
+};
+use light_registry::account_compression_cpi::sdk::{
+    create_batch_append_instruction, create_batch_nullify_instruction,
+};
+use light_utils::bigint::bigint_to_be_bytes_array;
+use light_verifier::CompressedProof;
+use reqwest::Client;
+use solana_program::pubkey::Pubkey;
+use solana_sdk::{signature::Keypair, signer::Signer};
+use tokio::sync::Mutex;
+use tracing::error;
+
+use crate::{errors::ForesterError, Result};
+
+pub struct BatchedOperations<R: RpcConnection, I: Indexer<R>> {
+    pub rpc_pool: Arc<SolanaRpcPool<R>>,
+    pub indexer: Arc<Mutex<I>>,
+    pub authority: Keypair,
+    pub derivation: Pubkey,
+    pub epoch: u64,
+    pub merkle_tree: Pubkey,
+    pub output_queue: Pubkey,
+}
+impl<R: RpcConnection, I: Indexer<R>> BatchedOperations<R, I> {
+    async fn is_batch_ready(&self) -> bool {
+        let mut rpc = self.rpc_pool.get_connection().await.unwrap();
+        let is_batch_ready = {
+            let mut output_queue_account =
+                rpc.get_account(self.output_queue).await.unwrap().unwrap();
+            let output_queue = BatchedQueueAccount::output_queue_from_bytes_mut(
+                output_queue_account.data.as_mut_slice(),
+            )
+            .unwrap();
+            output_queue.get_batch_num_inserted_in_current_batch() > 0
```

Comment:
I don't think this is a robust way to detect whether a batch or a part of a batch is ready for insertion.
This is the logic we use to check update readiness onchain, you can either reuse it and catch an error here or add a sub method which returns a boolean.
https://github.com/Lightprotocol/light-protocol/blob/3f44fbf094b64ccd82751bf9bbba84f0be7969d4/merkle-tree/batched/src/batch.rs#L114

Reply:
alternatively you can probably do:
```suggestion
            output_queue.get_batch_num_inserted_in_current_batch() % output_queue.zkp_batch_size == 0
```